graph/pull.go

```diff
@@ -641,20 +640,27 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
 		}(&downloads[i])
 	}
 
+	// Register the images serially. We must wait for all downloads to complete if one fails,
+	// in order to cleanup the others.
+	var lastErr error
 	var tagUpdated bool
 	for i := len(downloads) - 1; i >= 0; i-- {
-		d := &downloads[i]
-		if d.err != nil {
-			if err := <-d.err; err != nil {
-				return false, err
+		registerFunc := func(d *downloadInfo, pullAborted bool) error {
```

Comment:
I converted this into a function for three reasons:
1. to short-circuit the "pullAborted" case from trying to register
2. to invoke the deferred functions more quickly
3. in preparation for #14240 (i.e., fixing the FIXME below), where this will simplify the cleanup logic.